walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -26,19 +27,43 @@
     if (request.getDuration() > 1440)
       return Collections.emptyList();
 
+    boolean hasMandatoryAttendees = false;
+    boolean hasOptionalAttendees = false;
+
     SortedSet<TimeRange> mandatoryBlockers = new TreeSet<>(TimeRange.ORDER_BY_START);
+    SortedSet<TimeRange> optionalBlockers = new TreeSet<>(TimeRange.ORDER_BY_START);
 
     for (Event event : events) {
       for (String attendee : event.getAttendees()) {
         if (request.getAttendees().contains(attendee)) {
-          // this is a blocker, split up the available time segment
+          hasMandatoryAttendees = true;
           mandatoryBlockers.add(event.getWhen());
+          optionalBlockers.add(event.getWhen());
+        } else if (request.getOptionalAttendees().contains(attendee)) {
+          hasOptionalAttendees = true;
+          optionalBlockers.add(event.getWhen());
         }
       }
     }
 
     List<TimeRange> combinedMandatoryBlockers = mergeTimeRanges(mandatoryBlockers);
-    return subtractTimeRanges(combinedMandatoryBlockers, request.getDuration());
+    List<TimeRange> combinedOptionalBlockers = mergeTimeRanges(optionalBlockers);
+
+    if (hasMandatoryAttendees) {
```

Comment:
Instead of doing a deeply nested if-tree, you should take advantage of early returns:
```
if (!hasOptionalAttendees) {
  return onlyMandatoryTimes;
}

...
```